### ISSUE TITLE
Next13 hydration error when using suspense

### ISSUE THREAD

**Issue by Ziothh** [TanStack/query#5335]
### Describe the bug

I'm trying to use suspense but I'm getting a hydration error with this message:
"Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition."

I'm testing this in a new create-t3-app project, not using tRPC.

### Your minimal, reproducible example

https://github.com/Ziothh/rq-issue

### Steps to reproduce

This is my code:

```typescriptreact
import { useQuery } from "@tanstack/react-query"
import { FC, Suspense } from "react"

export default function page() {
  return (
    <div className="text-9xl">
      <Wrapper/>
    </div>
  )
}

const fetcher = async () => {
  return await new Promise<string>((res, rej) => {
    setTimeout(() => res("hi"), 4000)
  })
}

const Wrapper: FC = () => {
  return (
    <Suspense fallback={<h1>Loading....</h1>}>
      <PageContent/>
    </Suspense>
  )
}

const PageContent: FC = () => {
  const x = useQuery({
      queryKey: ["hi"], 
      queryFn: fetcher, 
       suspense: true,
   })

  return (
    <h1>{x.data ?? null}</h1>
  )
}
```

### Expected behavior

I'm expecting `PageContent` to suspend and `Wrapper` to render a fallback UI.

### How often does this bug happen?

Every time

### Platform

 - OS: MacOS
 - Next 13.2.4
 - Tanstack/react-query 4.28.0
 - React 18.2.0

### Tanstack Query adapter

react-query

### TanStack Query version

v4.28.0

### TypeScript version

v5.0.2

**Comment by TkDodo** [comment]
I can't open the reproduction in codesandbox because it keeps crashing :(

https://codesandbox.io/s/github/Ziothh/rq-issue

But, what I think is happening is that with React18, they started to implement streaming SSR, thus respecting suspense on the server. Previously, suspense on the server didn't do anything, and our queries didn't run on the server at all.

I don't think we are "there yet" with supporting suspsense on the server, so I think the best thing you can do is set `suspense: false` on the server and then only use it on the client.

@Ephem probably knows more

**Comment by TkDodo** [comment]
what I'm getting is:

```
❌ Invalid environment variables: {
  DATABASE_URL: [ 'Required' ],
  NEXTAUTH_URL: [ 'Required' ],
  DISCORD_CLIENT_ID: [ 'Required' ],
  DISCORD_CLIENT_SECRET: [ 'Required' ]
}
error - Failed to load next.config.mjs, see more info here https://nextjs.org/docs/messages/next-config-error
Error: Invalid environment variables
```

maybe you can take away all that unnecessary database and env dependencies from your reproduction? They seem totally unrelated...

**Comment by Ziothh** [comment]
Thanks for the quick response,

I have disabled the .env validation for the project.

Setting suspense: typeof window !== 'undefined' sadly does not work.

**Comment by TkDodo** [comment]
yeah you're right. My guess is react-query just doesn't work with suspense in react18 right now.

**Comment by TkDodo** [comment]
I wrote that comment, and I also added the `bug` and `suspense` label, so yeah ?

**Comment by TkDodo** [comment]
@Ziothh I would still appreciate a reproduction that is more minimal. no trpc, no t3, no prisma ... just adds lots of boilerplate to an otherwise easy to reproduce issue. Thanks.

**Comment by TkDodo** [comment]
Here is a minimal example that seems to work fine:

https://codesandbox.io/p/sandbox/dreamy-waterfall-2q98vx?file=%2Fpages%2Findex.tsx%3A7%2C27

**Comment by TkDodo** [comment]
So the culprit is actually `SessionProvider` from `next-auth`. This is EXACTLY the reason why a minimal example without other things is necessary to prove there's in issue in a certain library.

Here is your sandbox, session provider removed, and it works: https://codesandbox.io/s/rough-meadow-fi9no0

To the best of my knowledge, their code isn't safe for suspense streaming. I can see a bunch of `setState` in `useEffect`, and this probably happens while we're streaming things down, hence the error. They are not using `useSyncExternalStore`.

So yeah, I think I'm not getting back the time I spent (wasted) on this...
